File: frontend/netlify/functions/_shared/service.ts
import { coverage, difference, percentileRank, rangeStats, realPolicyRate, regime, REQUIRED_INDICATORS, round } from "./analytics";
import type { CatalogCountry, Country, Dataset, Observation } from "./types";

const CORE_INDICATORS = ["gdp_growth", "inflation", "policy_rate", "gov_10y", "current_account", "debt_gdp", "credit_growth"];

export class MacroService {
  private readonly catalog = new Map<string, CatalogCountry>();
  private readonly countries = new Map<string, Country>();
  private readonly series = new Map<string, Observation[]>();

  constructor(private readonly data: Dataset) {
    data.catalog.forEach((country) => this.catalog.set(country.code, country));
    data.countries.forEach((country) => this.countries.set(country.code, country));
    data.observations.forEach((row) => {
      const key = `${row.country_code}:${row.indicator_id}`;
      const values = this.series.get(key) || [];
      values.push(row);
      this.series.set(key, values);
    });
    this.series.forEach((rows) => rows.sort((a, b) => a.observation_date.localeCompare(b.observation_date) || a.period.localeCompare(b.period)));
  }

  metricSummary(code: string, indicatorId: string, period?: string | null) {
    const series = this.getSeries(code, indicatorId, period);
    if (!series.length) return emptyMetric(indicatorId);
    const latest = series.at(-1)!;
    const previous = series.at(-2);
    const values = series.map((row) => row.value);
    return {
      id: indicatorId,
      value: round(latest.value, 2),
      previous: previous ? round(previous.value, 2) : null,
      delta: previous ? difference(latest.value, previous.value) : null,
      percentile: percentileRank(values, latest.value),
      period: latest.period,
      observation_date: latest.observation_date,
      source: latest.source,
      validation_status: latest.validation_status,
      range: rangeStats(values),
      history: series.slice(-40).map((row) => ({ period: row.period, date: row.observation_date, value: round(row.value, 2) })),
    };
  }

  countrySnapshot(inputCode: string, period?: string | null) {
    const code = inputCode.toUpperCase();
    const catalog = this.catalog.get(code);
    const country = this.countries.get(code);
    if (!catalog && !country) throw new Error("NOT_FOUND");
    const metrics: Record<string, any> = Object.fromEntries(CORE_INDICATORS.map((id) => [id, this.metricSummary(code, id, period)]));
    const policyHistory = this.getSeries(code, "policy_rate", period);
    const inflationHistory = this.getSeries(code, "inflation", period);
    const realHistoryRows = policyHistory.flatMap((policy) => {
      const inflation = inflationHistory.filter((row) => row.observation_date <= policy.observation_date).at(-1);
      const value = realPolicyRate(policy.value, inflation?.value ?? null);
      return value == null ? [] : [{ policy, inflation, value }];
    });
    const realHistory = realHistoryRows.map((row) => row.value);
    const realValue = realPolicyRate(metrics.policy_rate.value, metrics.inflation.value);
    const previous = realHistory.at(-2) ?? null;
    metrics.real_policy_rate = {
      id: "real_policy_rate", value: realValue, previous,
      delta: difference(realValue, previous), percentile: percentileRank(realHistory, realValue),
      period: metrics.policy_rate.period, observation_date: metrics.policy_rate.observation_date,
      source: "Derived: policy rate - inflation", validation_status:
        metrics.policy_rate.validation_status === "VALID" && metrics.inflation.validation_status === "VALID" ? "VALID" : "FAILED",
      range: rangeStats(realHistory),
      history: realHistoryRows.slice(-40).map(({ policy, value }) => ({ period: policy.period, date: policy.observation_date, value })),
    };
    const base = country || {
      code, name: catalog!.name, region: catalog!.geographic_region,
      lat: catalog!.lat, lon: catalog!.lon, gdp_weight: null, tier: catalog!.tier, coverage_score: 0,
    };
    return {
      ...(catalog || {}), ...base,
      geographic_region: catalog?.geographic_region || base.region,
      ...coverage(metrics), country_code: code, country_name: base.name,
      latitude: base.lat, longitude: base.lon, metrics,
      regime: regime(metrics.gdp_growth.percentile, metrics.inflation.percentile),
      latest_period: metrics.gdp_growth.period,
    };
  }

  allSnapshots(period?: string | null) {
    return [...this.countries.keys()].map((code) => this.countrySnapshot(code, period));
  }

  universe(period?: string | null, universe = "all", tier?: number | null, region?: string | null, coverageMin = 0) {
    const snapshots = new Map(this.allSnapshots(period).map((country) => [country.code, country]));
    return [...this.catalog.values()].map((metadata) => snapshots.get(metadata.code) || this.countrySnapshot(metadata.code, period))
      .filter((country) => matches(country, universe, tier, region, coverageMin))
      .sort((a, b) => a.name.localeCompare(b.name));
  }

  coverageSummary(period?: string | null) {
    const countries = this.universe(period);
    const statuses = ["FULL", "GOOD", "PARTIAL", "LIMITED", "UNAVAILABLE"];
    return {
      total_economies: countries.length,
      analytical_economies: countries.filter((country) => country.available_indicator_count > 0).length,
      tiers: Object.fromEntries([1, 2, 3].map((tier) => [String(tier), countries.filter((country) => country.tier === tier).length])),
      distribution: Object.fromEntries(statuses.map((status) => [status, countries.filter((country) => country.data_status === status).length])),
      period: period || this.availablePeriods().at(-1) || null,
      required_indicators: [...REQUIRED_INDICATORS],
    };
  }

  regimeMap(period?: string | null) {
    return this.allSnapshots(period).filter((c) => c.metrics.gdp_growth.percentile != null && c.metrics.inflation.percentile != null)
      .map((c) => ({ code: c.code, name: c.name, growth_percentile: c.metrics.gdp_growth.percentile,
        inflation_percentile: c.metrics.inflation.percentile, regime: c.regime, gdp_weight: c.gdp_weight }));
  }

  policyMap(period?: string | null) {
    return this.allSnapshots(period).filter((c) => c.metrics.real_policy_rate.value != null)
      .map((c) => ({ code: c.code, name: c.name, policy_rate: c.metrics.policy_rate.value,
        inflation: c.metrics.inflation.value, real_rate: c.metrics.real_policy_rate.value,
        real_rate_percentile: c.metrics.real_policy_rate.percentile, gdp_weight: c.gdp_weight }))
      .sort((a, b) => b.real_rate - a.real_rate);
  }

  compare(left: string, right: string, period?: string | null) {
    const a = this.countrySnapshot(left, period);
    const b = this.countrySnapshot(right, period);
    const ids = ["gdp_growth", "inflation", "policy_rate", "real_policy_rate", "gov_10y", "current_account", "debt_gdp"];
    return { left: { code: a.code, name: a.name }, right: { code: b.code, name: b.name },
      rows: ids.map((indicator) => ({ indicator, left: a.metrics[indicator], right: b.metrics[indicator],
        spread: difference(a.metrics[indicator].value, b.metrics[indicator].value) })) };
  }

  availablePeriods() {
    return [...new Set(this.data.observations.map((row) => row.period))].sort();
  }

  timeline() {
    return this.availablePeriods().map((period) => ({ period, countries: this.regimeMap(period).map((c) => ({
      code: c.code, growth_percentile: c.growth_percentile, inflation_percentile: c.inflation_percentile, regime: c.regime,
    })) }));
  }

  private getSeries(code: string, indicator: string, period?: string | null) {
    const rows = this.series.get(`${code.toUpperCase()}:${indicator}`) || [];
    if (!period) return rows;
    const cutoff = periodEnd(period);
    return rows.filter((row) => row.observation_date <= cutoff);
  }
}

function periodEnd(period: string): string {
  const quarter = period.match(/^(\d{4})-Q([1-4])$/);
  if (quarter) return new Date(Date.UTC(Number(quarter[1]), Number(quarter[2]) * 3, 0)).toISOString().slice(0, 10);
  if (/^\d{4}-\d{2}$/.test(period)) {
    const [year, month] = period.split("-").map(Number);
    return new Date(Date.UTC(year, month, 0)).toISOString().slice(0, 10);
  }
  if (/^\d{4}$/.test(period)) return `${period}-12-31`;
  return period;
}

function emptyMetric(id: string) {
  return { id, value: null, previous: null, delta: null, percentile: null, period: null, history: [] };
}

function matches(country: any, universe: string, tier?: number | null, region?: string | null, coverageMin = 0) {
  if (tier != null && country.tier !== tier) return false;
  if (region && country.geographic_region.toLowerCase() !== region.toLowerCase()) return false;
  if (country.coverage_score < coverageMin) return false;
  const value = universe.toLowerCase();
  if (value === "all") return true;
  if (value === "core") return country.is_core;
  if (["g7", "g20", "asean"].includes(value)) return country.groups?.includes(value.toUpperCase());
  return country.geographic_region.toLowerCase() === value;
}

File: backend/app/service.py
from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timezone

from app.analytics.engine import change, percentile_rank, range_stats, real_policy_rate, regime, spread
from app.storage.repository import Repository
from app.coverage import coverage, matches, REQUIRED_INDICATORS

CORE_INDICATORS = [
    "gdp_growth",
    "inflation",
    "policy_rate",
    "gov_10y",
    "current_account",
    "debt_gdp",
    "credit_growth",
]


class MacroService:
    def __init__(self, repo: Repository | None = None):
        self.repo = repo or Repository()

    def metric_summary(self, code: str, indicator_id: str, period: str | None = None) -> dict:
        series = self.repo.indicator_series(code, indicator_id)
        if period:
            series = [row for row in series if row["period"] <= period]
        if not series:
            return {"id": indicator_id, "value": None, "previous": None, "delta": None, "percentile": None, "period": None, "history": []}
        values = [row["value"] for row in series]
        latest = series[-1]
        previous = series[-2] if len(series) > 1 else None
        return {
            "id": indicator_id,
            "value": round(latest["value"], 2),
            "previous": round(previous["value"], 2) if previous else None,
            "delta": change(latest["value"], previous["value"] if previous else None),
            "percentile": percentile_rank(values, latest["value"]),
            "period": latest["period"],
            "observation_date": latest["observation_date"],
            "source": latest["source"],
            "validation_status": latest["validation_status"],
            "range": range_stats(values),
            "history": [
                {"period": row["period"], "date": row["observation_date"], "value": round(row["value"], 2)}
                for row in series[-40:]
            ],
        }

    def country_snapshot(self, code: str, period: str | None = None) -> dict:
        country = self.repo.country(code)
        if not country:
            raise KeyError(code)
        metrics = {indicator: self.metric_summary(code, indicator, period) for indicator in CORE_INDICATORS}

        real_rate_value = real_policy_rate(metrics["policy_rate"]["value"], metrics["inflation"]["value"])
        policy_hist = self.repo.indicator_series(code, "policy_rate")
        inflation_hist = self.repo.indicator_series(code, "inflation")
        if period:
            policy_hist = [row for row in policy_hist if row["period"] <= period]
            inflation_hist = [row for row in inflation_hist if row["period"] <= period]
        n = min(len(policy_hist), len(inflation_hist))
        rr_history = [
            real_policy_rate(policy_hist[-n + i]["value"], inflation_hist[-n + i]["value"])
            for i in range(n)
        ] if n else []
        rr_percentile = percentile_rank(rr_history, real_rate_value)
        metrics["real_policy_rate"] = {
            "id": "real_policy_rate",
            "value": real_rate_value,
            "previous": rr_history[-2] if len(rr_history) > 1 else None,
            "delta": change(real_rate_value, rr_history[-2] if len(rr_history) > 1 else None),
            "percentile": rr_percentile,
            "period": metrics["policy_rate"]["period"],
            "observation_date": metrics["policy_rate"].get("observation_date"),
            "source": "Derived: policy rate - inflation",
            "validation_status": "VALID",
            "range": range_stats(rr_history),
            "history": [
                {"period": policy_hist[-n + i]["period"], "date": policy_hist[-n + i]["observation_date"], "value": rr_history[i]}
                for i in range(n)
            ][-40:],
        }

        macro_regime = regime(metrics["gdp_growth"]["percentile"], metrics["inflation"]["percentile"])
        return {
            **self.repo.metadata(code),
            **country,
            "geographic_region": self.repo.metadata(code).get("region", country["region"]),
            **coverage(metrics),
            "country_code": country["code"], "country_name": country["name"],
            "latitude": country["lat"], "longitude": country["lon"],
            "gdp_weight": country.get("gdp_weight"),
            "metrics": metrics,
            "regime": macro_regime,
            "latest_period": metrics["gdp_growth"]["period"],
        }

    def all_snapshots(self, period: str | None = None) -> list[dict]:
        return [self.country_snapshot(c["code"], period) for c in self.repo.countries()]

    def universe(self, period: str | None = None, universe: str = "all", tier: int | None = None,
                 region: str | None = None, coverage_min: float = 0) -> list[dict]:
        # Only query series for economies present in the analytical repository.
        observed = {c["code"]: self.country_snapshot(c["code"], period) for c in self.repo.countries()}
        empty_metrics = {i: {"id": i, "value": None, "previous": None, "delta": None,
                            "percentile": None, "period": None, "history": []} for i in REQUIRED_INDICATORS}
        output = []
        catalog = {c["code"]: c for c in self.repo.catalog()}
        for code, metadata in {**catalog, **{c: catalog.get(c, {}) for c in observed}}.items():
            snapshot = observed.get(code) or {**metadata, **coverage(empty_metrics), "metrics": empty_metrics,
                "country_code": code, "country_name": metadata["name"], "latitude": metadata["lat"],
                "longitude": metadata["lon"], "gdp_weight": None, "regime": "UNCLASSIFIED", "latest_period": None}
            if matches(snapshot, universe, tier, region, coverage_min):
                output.append(snapshot)
        return sorted(output, key=lambda c: c["name"])

    def coverage_summary(self, period: str | None = None) -> dict:
        countries = self.universe(period)
        return {"total_economies": len(countries),
                "analytical_economies": sum(c["available_indicator_count"] > 0 for c in countries),
                "tiers": {str(t): sum(c["tier"] == t for c in countries) for t in (1, 2, 3)},
                "distribution": {s: sum(c["data_status"] == s for c in countries)
                                 for s in ("FULL", "GOOD", "PARTIAL", "LIMITED", "UNAVAILABLE")},
                "period": period or self.repo.latest_period(),
                "required_indicators": list(REQUIRED_INDICATORS)}

    def regime_map(self, period: str | None = None) -> list[dict]:
        items = []
        for snapshot in self.all_snapshots(period):
            if snapshot["metrics"]["gdp_growth"]["percentile"] is None or snapshot["metrics"]["inflation"]["percentile"] is None:
                continue
            items.append({
                "code": snapshot["code"],
                "name": snapshot["name"],
                "growth_percentile": snapshot["metrics"]["gdp_growth"]["percentile"],
                "inflation_percentile": snapshot["metrics"]["inflation"]["percentile"],
                "regime": snapshot["regime"],
                "gdp_weight": snapshot["gdp_weight"],
            })
        return items

    def policy_map(self, period: str | None = None) -> list[dict]:
        items = []
        for snapshot in self.all_snapshots(period):
            if snapshot["metrics"]["real_policy_rate"]["value"] is None:
                continue
            items.append({
                "code": snapshot["code"],
                "name": snapshot["name"],
                "policy_rate": snapshot["metrics"]["policy_rate"]["value"],
                "inflation": snapshot["metrics"]["inflation"]["value"],
                "real_rate": snapshot["metrics"]["real_policy_rate"]["value"],
                "real_rate_percentile": snapshot["metrics"]["real_policy_rate"]["percentile"],
                "gdp_weight": snapshot["gdp_weight"],
            })
        return sorted(items, key=lambda item: item["real_rate"] if item["real_rate"] is not None else -999, reverse=True)

    def compare(self, left: str, right: str, period: str | None = None) -> dict:
        a = self.country_snapshot(left, period)
        b = self.country_snapshot(right, period)
        ids = ["gdp_growth", "inflation", "policy_rate", "real_policy_rate", "gov_10y", "current_account", "debt_gdp"]
        rows = []
        for indicator in ids:
            am = a["metrics"][indicator]
            bm = b["metrics"][indicator]
            rows.append({
                "indicator": indicator,
                "left": am,
                "right": bm,
                "spread": spread(am["value"], bm["value"]),
            })
        return {"left": {"code": a["code"], "name": a["name"]}, "right": {"code": b["code"], "name": b["name"]}, "rows": rows}

    def snapshots_by_period(self) -> list[dict]:
        periods = self.repo.available_periods()
        output: list[dict] = []
        for period in periods:
            countries: list[dict] = []
            for country in self.repo.countries():
                code = country["code"]
                gp_series = [r for r in self.repo.indicator_series(code, "gdp_growth") if r["period"] <= period]
                inf_series = [r for r in self.repo.indicator_series(code, "inflation") if r["period"] <= period]
                if not gp_series or not inf_series:
                    continue
                gp = gp_series[-1]
                inf = inf_series[-1]
                gp_pct = percentile_rank([r["value"] for r in gp_series], gp["value"])
                inf_pct = percentile_rank([r["value"] for r in inf_series], inf["value"])
                countries.append({
                    "code": code,
                    "growth_percentile": gp_pct,
                    "inflation_percentile": inf_pct,
                    "regime": regime(gp_pct, inf_pct),
                })
            output.append({"period": period, "countries": countries})
        return output

    def system_status(self) -> dict:
        countries = len(self.repo.countries())
        with self.repo.connect() as conn:
            series_count = conn.execute("SELECT COUNT(DISTINCT country_code || ':' || indicator_id) AS c FROM observations").fetchone()["c"]
            observations = conn.execute("SELECT COUNT(*) AS c FROM observations").fetchone()["c"]
            warnings = conn.execute("SELECT COUNT(*) AS c FROM observations WHERE validation_status != 'VALID'").fetchone()["c"]
            last = conn.execute("SELECT MAX(retrieved_at) AS d FROM observations").fetchone()["d"]
        return {
            "mode": "DEMO_SNAPSHOT",
            "countries": countries,
            "series": series_count,
            "observations": observations,
            "validation_warnings": warnings,
            "last_sync": last,
            "latest_period": self.repo.latest_period(),
            "data_label": "DEMO DATA — connect official APIs before production use",
            "generated_at": datetime.now(timezone.utc).isoformat(),
        }

    def lineage(self, code: str, indicator_id: str) -> dict:
        indicator = self.repo.indicator(indicator_id)
        latest = self.repo.latest_observation(code, indicator_id)
        if not indicator or not latest:
            raise KeyError(f"{code}:{indicator_id}")
        summary = self.metric_summary(code, indicator_id)
        return {
            "country": self.repo.country(code),
            "indicator": indicator,
            "observation": latest,
            "derived": {
                "historical_percentile": summary["percentile"],
                "historical_window": "demo history / trailing available observations",
            },
            "methodology": "Observed data + deterministic transformations. No forecasting or causal estimation.",
        }
